### DISCUSSION TITLE
support for virtual columns

### DISCUSSION THREAD

**Discussion by michalfairfly** [dbt-labs/dbt-snowflake#960] [Ideas]
Hello, i have created for my internal needs support for virtual columns,
it could be useful for having access to simple automated calculations without need to store the column or remember the logic behind them (or creating a view):
Datetime columns in different timezones, for example `arrival`, `arrival_in_utc`, `arrival_in_local_time_zone`, `arrival_in_departure_timezone`  

this type of construct will not affect incremental runs.

in file [create.sql](dbt/include/snowflake/macros/relations/table/create.sql) add following
```jinja
    {%- set virtual_columns = config.get('virtual_columns', default=none) -%}

...
 -- add before clustering logic to allow for clustering over virtual column 
  {% if virtual_columns is not none %}
        {% for key, value in virtual_columns.items() %}
          alter table {{ relation }} add column {{ key }} {{ value.type }} as ({{ value.expression }});
        {% endfor %}
      {% endif %}

```

in [impl.py](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/adapters/snowflake/impl.py)
add to the SnowflakeConfig
 ```python
@dataclass
class SnowflakeConfig(AdapterConfig):
    transient: Optional[bool] = None
    cluster_by: Optional[Union[str, List[str]]] = None
    automatic_clustering: Optional[bool] = None
    secure: Optional[bool] = None
    copy_grants: Optional[bool] = None
    snowflake_warehouse: Optional[str] = None
    query_tag: Optional[str] = None
    tmp_relation_type: Optional[str] = None
    merge_update_columns: Optional[str] = None
    target_lag: Optional[str] = None
    virtual_columns: Optional[Dict[str, Any]] = None
```

and then the config for snowflake model could look like as following:
```
{{
    config(
        materialized='incremental',
        unique_key='incremental_id',
        virtual_columns={
            'some_virtual_field': {"type": "float", "expression": "price* rate"},
            'another_virtual_field': {"type": "number(38,0)", "expression": "id + 1"}
        }
    )
}}
```

**Comment by dataders** [comment]
am I way off base in wanting to draw connections to both https://github.com/dbt-labs/dbt-external-tables/issues/140 and https://github.com/dbt-labs/dbt-core/issues/6929? In some sense a virtual column is a column alias that's kind of like a view materialization?
